allow server selection to be aware of query

kroeders opened the pull request:
### Description

This PR supports https://github.com/apache/druid/issues/10294 where an issue is discussed that queries with lookups may sometimes fail unnecessarily due to the lookup only being available on some historicals. The query can succeed by avoiding those historicals. In this PR is an extension that tracks lookup status on the broker and adds a new ServerSelectorStrategy to avoid historicals that are missing required lookups. 

This PR https://github.com/apache/druid/pull/10427 is a small change to make server selection aware of the Query and would allow this extension so solve this issue without any other changes in core Druid. 

This change just adds Query as a parameter to the ServerSelectorStratey and TierSelectorStrategy and provides this parameter in CachingClusteredClient, after segments  are found for a query and they are being fanned out across the cluster. A default method was added to the interface so existing implementations would not need to change. 

<hr>

This PR has:
- [X] been self-reviewed.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [X] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `ServerSelectorStrategy`
 * `TierSelectorStrategy`
 * `CachingClusteredClient`

a2l007 reviewed (approved):
+1 after CI

kroeders commented:
thanks for taking a look @nishantmonu51 !

jihoonson commented on server/src/main/java/org/apache/druid/client/selector/TierSelectorStrategy.java:
```diff
@@ -47,6 +49,22 @@
   List<QueryableDruidServer> pick(
```
Any reason to keep the old interfaces? If this is for backward-compatibility, it maybe make sense. But please annotate them as `@Deprecated`.

kroeders replied:
I marked the non-query ones as deprecated throughout

jihoonson commented on server/src/main/java/org/apache/druid/client/selector/TierSelectorStrategy.java:
```diff
@@ -47,6 +49,22 @@
+      Int2ObjectRBTreeMap<Set<QueryableDruidServer>> prioritizedServers,
+      DataSegment segment)
+  {
+    return pick(prioritizedServers, segment);
```
If you want to keep old interfaces, I think the old ones should have a default implementation that calls this new interface with `null` query. Then we can remove the implementations of old interfaces in `AbstractTierSelectorStrategy`.

kroeders replied:
This was done for backward compatibility - I was trying to make this change as minimal as possible on the first cut. We don't have any custom implementations here and I haven't seen any on a cursory look, but I don't want to presume. I think it's cleaner to remove the ones that don't take Query and having an interface of all defaults that call each other makes the abstract methods less obvious. What do you think? I'm happy to defer to your judgement.

jihoonson replied:
I don't have strong opinion here. The current one looks good to me.

kroeders replied:
works for me, current one it is!

jihoonson commented on server/src/main/java/org/apache/druid/client/selector/TierSelectorStrategy.java:
```diff
@@ -47,6 +49,22 @@
+      int numServersToPick);
+
+  @Nullable
+  default <T> QueryableDruidServer pick(Query<T> query,
```
Please annotate query with `@Nullable`.

kroeders replied:
done

jihoonson commented on server/src/main/java/org/apache/druid/client/selector/TierSelectorStrategy.java:
```diff
@@ -47,6 +49,22 @@
+  }
+
+  default <T> List<QueryableDruidServer> pick(
+      Query<T> query,
```
Please add `@Nullable`.

kroeders replied:
done

jihoonson commented on server/src/main/java/org/apache/druid/client/selector/ServerSelectorStrategy.java:
```diff
@@ -33,6 +35,17 @@
 })
 public interface ServerSelectorStrategy
 {
+  default <T> QueryableDruidServer pick(Query<T> query, Set<QueryableDruidServer> servers, DataSegment segment)
```
Similarly, I think it would be better that old interfaces have a default implementation that calls new interfaces with `null` query. Also please annotate `query` with `@Nullable`.

kroeders replied:
done

jihoonson commented on server/src/main/java/org/apache/druid/client/selector/ServerSelector.java:
```diff
@@ -162,12 +163,20 @@ public boolean isEmpty()
+    return strategy.pick(realtimeServers, segment.get());
+  }
+
+  public <T> QueryableDruidServer pick(Query<T> query)
```
It seems that `pick()` is no longer in use. I think we can remove `pick()` but keep only this new one if `ServerSelector` doesn't implement `DiscoverySelector`.

kroeders replied:
I removed it, the method was called a couple of times in tests (which seems OK) and in the etag creation part, which seems OK to me. It looks a little funny to pass in null as a single parameter, but I could argue either way vs having a method that isn't used elsewhere.

jihoonson reviewed (commented):
@kroeders sorry for delayed review. I like the idea overall, it seems pretty extendable that we can add more functionalities/optimizations in the future as serverSelector is query-aware. I left some comments around default methods of new interfaces.

kroeders commented:
thanks for the feedback and taking the time to look @jihoonson , I appreciate it! I updated the branch per the comments and left a few notes that I found while making the changes.

kroeders commented on server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:
```diff
@@ -812,7 +812,7 @@ String computeResultLevelCachingEtag(
       boolean hasOnlyHistoricalSegments = true;
       for (SegmentServerSelector p : segments) {
-        if (!p.getServer().pick().getServer().isSegmentReplicationTarget()) {
+        if (!p.getServer().pick(null).getServer().isSegmentReplicationTarget()) {
```
this was the only functional place that the parameterless pick() was called that I mentioned in the other comment @jihoonson , the Query is available inside of SpecificQueryRunnable where this is called, so I could alternatively pass the query into computeResultLevelCachingEtag and add a null check on this line, what do you think?

kroeders replied:
I made the query change.
For the null check I was thinking for this part p.getServer().pick(query) since we just annotated this Nullable below?
Something like -
QueryableServer queryableServer = p.getServer().pick(query);
if (queryableServer==null || !queryableServer.getServer().isSegmentReplicationTarget()) {

jihoonson replied:
That's a good catch! I think we should add a null check.

jihoonson reviewed (commented):
@kroeders thanks for addressing comments. I left some trivial comments. Other than those, this PR LGTM. I will approve once those comments are addressed.

kroeders commented on server/src/main/java/org/apache/druid/client/selector/ServerSelector.java:
```diff
@@ -159,15 +160,13 @@ public boolean isEmpty()
-  @Nullable
-  @Override
-  public QueryableDruidServer pick()
+  public <T> QueryableDruidServer pick(@Nullable Query<T> query)
```
yep, it sure can! thanks

kroeders commented:
fixing a few test issues with the default methods...

kroeders commented:
hi @jihoonson I think I've got them all here. I also added@Nullable to the ServerSelectorStrategy method that leads to this other Nullable and added a couple tests. Please let me know if you see anything else and thanks again for working with me on this one

jihoonson reviewed (commented):
@kroeders sorry, I got side-tracked for a while. The latest changes LGTM, except for the null check you found. It would be great to fix it here too.

kroeders commented:
hi @jihoonson no worries of course, thanks for working with me on this! This will let us package the lookup change as an extension and deploy it internally. I added the null check and pushed here, happy holidays

jihoonson commented on server/src/main/java/org/apache/druid/client/CachingClusteredClient.java:
```diff
@@ -812,7 +812,8 @@ String computeResultLevelCachingEtag(
       for (SegmentServerSelector p : segments) {
-        if (!p.getServer().pick().getServer().isSegmentReplicationTarget()) {
+        QueryableDruidServer queryableServer = p.getServer().pick(query);
+        if (queryableServer == null || !queryableServer.getServer().isSegmentReplicationTarget()) {
```
nit: I think it should ideally fail if `queryableServer` is null, because it means there is no server to process this segment. However, this seems OK because this method will return immediately when it's null and fail at [here](https://github.com/apache/druid/pull/10428/files#diff-d0873a0b1d76e92318ed707d876a168cf8ad974a4c53ebe65d7a39ab3cf0294eR594).

jihoonson reviewed (approved):
@kroeders thanks for fixing it quickly! +1 after Travis passes.

kroeders commented:
@jihoonson no problem, thanks again for working with me on this one!

jihoonson commented:
@kroeders I just merged. Thank you!

kroeders commented:
awesome, thanks!
